Add getters and watchers for oneSignalId and externalId

jinliu9508 opened the pull request:
# Description
## One Line Summary
Add getters and observer for onesignalId and externalId.

## Details

### Motivation
Exposing the onesignal ID and external ID with getters.
Also grant developers the ability to add observers that can be called when there is a change in user state.

### Scope
UserState contains onesignalId and externalId (both can be empty strings if not available or not set).

# Testing
## Unit testing
No unit testing at this time

## Manual testing
I have added two testing buttons in the test scene for getting onesignalId and externalId. I have also add observer to test the callback when the ids are changed.

Unity version: 2022.3.10f1
For Android: Emulator Pixel 3a api34
For ios: iPhone15 Simulator

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [ ] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-Unity-SDK/690)
<!-- Reviewable:end -->

shepherd-l commented on MIGRATION_GUIDE_v3_to_v5.md:
```diff
@@ -99,6 +99,14 @@ Once (or if) the user is no longer identifiable in your app (i.e. they logged ou
 
+To observe changes to the onesignalId or externalId you can add a custom method to the event:
+    
+    OneSignal.User.Changed += _userStateChanged;
+
+    private void yourOnUserStateChangedMethod(object sender, UserStateChangedEventArgs e) {
```
I think the method names should match in the example to avoid confusion

shepherd-l commented on MIGRATION_GUIDE_v3_to_v5.md:
```diff
@@ -165,7 +173,10 @@ The user name space is accessible via `OneSignal.User` and provides access to us
 | ----------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `IPushSubscription PushSubscription`                  | *The push subscription associated to the current user.*                                                                                                                                                                                  |
 | `string Language`                                     | *Set the 2-character language either as a detected language or explicitly set for this user.*                                                                                                                                            |
```
Can you also add the OneSignalId and ExternalId properties in the table as well?

shepherd-l commented on OneSignalExample/Assets/OneSignal/CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - iOS crash when calling OneSignal.User.PushSubscription.Id and OneSignal.User.PushSubscription.Token when they are null.
 
+## [5.1.0]
```
The release script will add this line so this doesn't need to be in this PR

shepherd-l commented on com.onesignal.unity.android/Runtime/AndroidUserManager.cs:
```diff
@@ -93,7 +104,32 @@ public void RemoveSms(string sms)
+
+            /// <param name="state">UserChangedState</param>
+            public void onUserStateChange(AndroidJavaObject state) {
+                UnityEngine.Debug.Log("onUserStateChange fired");
```
I don't think this debug log is needed since the other events don't have logs. We could add debug logs to all the other events instead. What do you think?

shepherd-l commented on com.onesignal.unity.core/Runtime/User/IUserManager.cs:
```diff
@@ -46,6 +61,24 @@ namespace OneSignalSDK.User {
+        /// <summary>
+        /// The UUID generated by OneSignal to represent a user, empty if this is currently unavailable.
+        /// </summary>
+        string onesignalId { get; }
```
Can you change OneSignalId to be Pascal case? To match the coding style

shepherd-l commented on com.onesignal.unity.core/Runtime/User/IUserManager.cs:
```diff
@@ -46,6 +61,24 @@ namespace OneSignalSDK.User {
+        ///
+        /// This is empty if the External ID has not been set.
+        /// </summary>
+        string externalId { get; }
```
Change ExternalId to Pascal case too

shepherd-l commented on com.onesignal.unity.core/Runtime/User/IUserState.cs:
What do you think about putting this file in the Models folder similar to IPushSubcription.cs?

shepherd-l commented on com.onesignal.unity.core/Runtime/User/UserChangedState.cs:
What do you think about putting this file in the Models folder similar to PushSubscriptionChangedState.cs?

shepherd-l commented on OneSignalExample/Assets/OneSignal/CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## [5.1.0]
+### Fixed
+- Add getters for the user state that includes onesignalId and externalId
```
I think you can just put something like "Add public getters for OnesignalId and ExternalId in the User namespace" to make things simpler

I think this should be categorized under ###Changed because adding public API is a change and not a fix

shepherd-l commented on OneSignalExample/Assets/OneSignal/CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+## [5.1.0]
+### Fixed
+- Add getters for the user state that includes onesignalId and externalId
+- Add observer for the user state
```
I think we could also specify the name of the event handler. Something like, "Add public event handler OneSignal.User.Changed that fires when the OnesignalId or ExternalId changes"

I also think this should be categorized under ###Changed

shepherd-l reviewed (changes requested):
Just had some nits, looks good!

Could you also update the Manual testing section in your PR to include what devices and version of Unity you tested on? I think it would be helpful to have it documented

WIP: I will test the PR on devices later

jinliu9508 commented:
> Just had some nits, looks good!
> 
> Could you also update the Manual testing section in your PR to include what devices and version of Unity you tested on? I think it would be helpful to have it documented
> 
> WIP: I will test the PR on devices later

@shepherd-l Thanks for the review! I have made changes addressing all the concerns. Please have a look again.

shepherd-l reviewed (approved):
LGTM!
Tested new install, login, and logout on an emulated Pixel 4 with Android 12 and physical iPhone 12 with iOS 15.5 built with Unity 2022.3.10f1

shepherd-l commented on MIGRATION_GUIDE_v3_to_v5.md:
```diff
@@ -165,7 +173,12 @@ The user name space is accessible via `OneSignal.User` and provides access to us
 | ----------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `IPushSubscription PushSubscription`                  | *The push subscription associated to the current user.*                                                                                                                                                                                  |
 | `string Language`                                     | *Set the 2-character language either as a detected language or explicitly set for this user.*                                                                                                                                            |
+| `string onesignalId`                                  | *The UUID generated by OneSignal to represent a user, empty if this is currently unavailable.*                                                                                                                                            |
```
I think this should be OneSignalId

shepherd-l commented on MIGRATION_GUIDE_v3_to_v5.md:
```diff
@@ -165,7 +173,12 @@ The user name space is accessible via `OneSignal.User` and provides access to us
 | `IPushSubscription PushSubscription`                  | *The push subscription associated to the current user.*                                                                                                                                                                                  |
 | `string Language`                                     | *Set the 2-character language either as a detected language or explicitly set for this user.*                                                                                                                                            |
+| `string onesignalId`                                  | *The UUID generated by OneSignal to represent a user, empty if this is currently unavailable.*                                                                                                                                            |
+| `string externalId`                                   | *The External ID is OneSignal's default and recommended alias label. This should be the mainidentifier you use to identify users. It is set when calling the [OneSignal.login] method.*                                                                                                                                            |
```
I think this should be ExternalId

shepherd-l commented on com.onesignal.unity.core/Samples~/OneSignalExampleBehaviour.cs:
```diff
@@ -211,6 +212,11 @@ public class OneSignalExampleBehaviour : MonoBehaviour {
     }
 
+    private void _userStateChanged(object sender, UserStateChangedEventArgs e) {
+        _log($"onesignalId changed : {JsonUtility.ToJson(e.State.Current.onesignalId)}");
```
I think this should be updated to OneSignalId

shepherd-l commented on com.onesignal.unity.core/Samples~/OneSignalExampleBehaviour.cs:
```diff
@@ -211,6 +212,11 @@ public class OneSignalExampleBehaviour : MonoBehaviour {
 
+    private void _userStateChanged(object sender, UserStateChangedEventArgs e) {
+        _log($"onesignalId changed : {JsonUtility.ToJson(e.State.Current.onesignalId)}");
+        _log($"externalId changed : {JsonUtility.ToJson(e.State.Current.externalId)}");
```
I think this should be updated to ExternalId

shepherd-l commented on MIGRATION_GUIDE_v3_to_v5.md:
```diff
@@ -99,6 +99,14 @@ Once (or if) the user is no longer identifiable in your app (i.e. they logged ou
 
 Logging out has the affect of reverting to a “device-scoped” user, which is the new owner of the device’s push subscription.
 
+To observe changes to the onesignalId or externalId you can add a custom method to the event:
```
I think "onesignalId" could be changed to "OneSignal Id" and "externalId" to "external Id" to avoid it being confused for variables

shepherd-l commented on com.onesignal.unity.core/Runtime/User/Models/IUserState.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// The unique identifier for your OneSignal account. This will be an empty string until the
+        /// user has been successfully logged in on the backend and assigned an ID.
+        /// Use [addObserver] to be notified when the [OnesignalId] has been successfully assigned.
```
I think addObserver is language from the native SDKs. Maybe change [addObserver] to "[Changed] event handler"

shepherd-l commented on com.onesignal.unity.core/Runtime/User/Models/IUserState.cs:
```diff
@@ -0,0 +1,46 @@
+        string OneSignalId { get; }
+        
+        /// <summary>
+        /// The external identifier that you use to identify users. Use [addObserver] to be notified
+        /// when the [externalId] has been successfully assigned. This will be an empty string if no
```
Same as above for addObserver to Changed Event Handler

Also update [externalId] to ExternalId to match property

shepherd-l reviewed (changes requested):
Some more nits

shepherd-l reviewed (approved):
LGTM!
